src/hotspot/os/linux/memMapPrinter_linux.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Red Hat, Inc. and/or its affiliates.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "runtime/os.hpp"
+#include "nmt/memMapPrinter.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include <limits.h>
+
+struct ProcMapsInfo {
+  void* from = 0;
+  void* to = 0;
+  char prot[20 + 1];
+  char offset[20 + 1];
+  char dev[20 + 1];
+  char inode[20 + 1];
+  char filename[1024 + 1];
+
+  bool scan_proc_maps_line(const char* line) {
+    prot[0] = offset[0] = dev[0] = inode[0] = filename[0] = '\0';
+    const int items_read = ::sscanf(line, "%p-%p %20s %20s %20s %20s %1024s",
+        &from, &to, prot, offset, dev, inode, filename);
+    return items_read >= 2; // need at least from and to
+  }
+};
+
+class LinuxMappingPrintInformation : public MappingPrintInformation {
+  const ProcMapsInfo _info;
+public:
+
+  LinuxMappingPrintInformation(const void* from, const void* to, const ProcMapsInfo* info) :
+    MappingPrintInformation(from, to), _info(*info) {}
+
+  void print_OS_specific_details(outputStream* st) const override {
+    st->print("%s %s ", _info.prot, _info.offset);
+  }
```

Comment:
This prints protection and offset. The former is interesting for obvious reasons (e.g. lets you tell apart stack guard regions from stack, or uncommitted from committed regions) and the latter interesting for some corner cases (e.g. ZGC mapped regions).

I don't want to do this in shared code though since I am not sure every OS has this information or whether I can obtain this information. I rather not print out a bunch of 0s or empty strings.

The alternative would be carrying a non-descriptive text string around with "OS information", but I don't want to have to think about storage and copying stuff around. It would not be simpler than this.

BTW, the way to implement this on MacOS would be probably by spawning off `vmmap` and parsing the output; on Windows one would use `VirtualQuery`. But I leave this for other folks, I am happy with Linux for now.

Reply:
On ZGC, by the way, it's possible to see many `memfd:java_heap (deleted)`, does this work allow annotate these mappings ?

Reply:
Yes, they should show up as JAVAHEAP.